### PR TITLE
Add batch job to initialize the webhook secret to manifest, remove requirement of manual step

### PR DESCRIPTION
**Pull request by mrow4a**
CC @liyinan926

### CODE REVIEW

**Comment by mrow4a** [comment]
I think we can at some point get rid of all manifests, yes. But I think in alpha it is still useful.

**Comment by liyinan926** [comment]
Instead of using a `Job`, can we use an init-container to generate the secret? This guarantees that it gets done before the operator starts running.

**Comment by liyinan926** [comment]
What if the secret has already been created? For example, if the user updates the operator Deployment spec and does `kubectl apply -f ...`. In this case, `gencerts.sh` will fail creating the secret. Will this error cause the Job to be restarted? Another question is is it guaranteed that the `Job` is created and run before the operator Deployment is created?

**Comment by mrow4a** [comment]
Spark Operator waits on container creation for secret to be mounted. The same is done in Helm to simplify it. If secret exists the job will just terminate with error.

**Comment by liyinan926** [comment]
> Spark Operator waits on container creation for secret to be mounted. The same is done in Helm to simplify it. If secret exists the job will just terminate with error.

The `OnFailure` restart policy of the job may cause it to restart. I'm actually thinking about augmenting `sparkctl` with an `install` subcommand for installing the operator with options for installing the webhook.

File: hack/gencerts.sh [hunk 1]
```diff
@@ -127,7 +131,22 @@ if [ "$IN_POD" == "true" ];  then
...
+        printf "\nSuccess - secret created.\n"
+      ;;
+      409)
+        printf "\nSuccess - secret already exists.\n"
+      ;;
+      *)
+        printf "\nFailed creating secret.\n"
+        exit 1
+      ;;
+    esac
```
Comment (mrow4a) [inline hack/gencerts.sh]: @liyinan926 @yuchaoran2011 now I explicitly fail on any error like 
```
Creating a secret for the certificate and keys
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  4  8033  100   329    0     0   1920      0 --:--:-- --:--:-- --:--:--  1912
HTTP/1.1 403 Forbidden
Content-Type: application/json
X-Content-Type-Options: nosniff
Date: Fri, 21 Sep 2018 13:28:41 GMT
Content-Length: 329
Connection: close

{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "secrets is forbidden: User \"system:serviceaccount:sparkoperator:sparkoperator\" cannot create secrets in the namespace \"sparkoperator\"",
  "reason": "Forbidden",
  "details": {
    "kind": "secrets"
  },
  "code": 403
}
Failed creating secret.
```

and succeed with 
```
HTTP/1.1 409 Conflict
Content-Type: application/json
Date: Fri, 21 Sep 2018 13:31:08 GMT
Content-Length: 277

{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "secrets \"spark-webhook-certs\" already exists",
  "reason": "AlreadyExists",
  "details": {
    "name": "spark-webhook-certs",
    "kind": "secrets"
  },
  "code": 409
}
Success - secret already exists.
```

or 

```
HTTP/1.1 201 Created
Content-Type: application/json
Date: Fri, 21 Sep 2018 13:32:32 GMT
Transfer-Encoding: chunked

{
  "kind": "Secret",
  "apiVersion": "v1",
  "metadata": {
    "name": "spark-webhook-certs",
    "namespace": "sparkoperator",
    "selfLink": "/api/v1/namespaces/sparkoperator/secrets/spark-webhook-certs",
    "uid": "redacted",
    "resourceVersion": "2384255",
    "creationTimestamp": "2018-09-21T13:32:32Z"
  },
  "data": {
    "ca-cert.pem": "redacted",
    "ca-key.pem": "redacted",
    "server-cert.pem": "redacted",
    "server-key.pem": "redacted"
  },
  "type": "Opaque"
}
Success - secret created.
```

File: manifest/spark-operator-with-webhook.yaml [hunk 2]
```diff
@@ -57,6 +57,30 @@ spec:
...
+spec:
+  backoffLimit: 3
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: sparkoperator
+        app.kubernetes.io/version: v2.3.1-v1alpha1
+    spec:
+      serviceAccountName: sparkoperator
+      restartPolicy: Never
```
Comment (mrow4a) [inline manifest/spark-operator-with-webhook.yaml]: I never restart container

File: manifest/spark-operator-with-webhook.yaml [hunk 3]
```diff
@@ -57,6 +57,30 @@ spec:
...
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: sparkoperator-init
+  namespace: sparkoperator
+  labels:
+    app.kubernetes.io/name: sparkoperator
+    app.kubernetes.io/version: v2.3.1-v1alpha1
+spec:
+  backoffLimit: 3
```
Comment (mrow4a) [inline manifest/spark-operator-with-webhook.yaml]: Allows to try recreating pods max 3 times

File: hack/gencerts.sh [hunk 4]
```diff
@@ -127,7 +131,22 @@ if [ "$IN_POD" == "true" ];  then
...
+	case "$STATUS" in
+	  201)
+	    printf "\nSuccess - secret created.\n"
+	  ;;
+	  409)
+	    printf "\nSuccess - secret already exists.\n"
+	  ;;
+	  *)
+	    printf "\nFailed creating secret.\n"
+	    exit 1
```
Comment (mrow4a) [inline hack/gencerts.sh]: @liyinan926 @yuchaoran2011 now I explicitly fail on any error like 
```
Creating a secret for the certificate and keys
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  4  8033  100   329    0     0   1920      0 --:--:-- --:--:-- --:--:--  1912
HTTP/1.1 403 Forbidden
Content-Type: application/json
X-Content-Type-Options: nosniff
Date: Fri, 21 Sep 2018 13:28:41 GMT
Content-Length: 329
Connection: close

{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "secrets is forbidden: User \"system:serviceaccount:sparkoperator:sparkoperator\" cannot create secrets in the namespace \"sparkoperator\"",
  "reason": "Forbidden",
  "details": {
    "kind": "secrets"
  },
  "code": 403
}
Failed creating secret.
```

and succeed with 
```
HTTP/1.1 409 Conflict
Content-Type: application/json
Date: Fri, 21 Sep 2018 13:31:08 GMT
Content-Length: 277

{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "secrets \"spark-webhook-certs\" already exists",
  "reason": "AlreadyExists",
  "details": {
    "name": "spark-webhook-certs",
    "kind": "secrets"
  },
  "code": 409
}
Success - secret already exists.
```

or 

```
HTTP/1.1 201 Created
Content-Type: application/json
Date: Fri, 21 Sep 2018 13:32:32 GMT
Transfer-Encoding: chunked

{
  "kind": "Secret",
  "apiVersion": "v1",
  "metadata": {
    "name": "spark-webhook-certs",
    "namespace": "sparkoperator",
    "selfLink": "/api/v1/namespaces/sparkoperator/secrets/spark-webhook-certs",
    "uid": "redacted",
    "resourceVersion": "2384255",
    "creationTimestamp": "2018-09-21T13:32:32Z"
  },
  "data": {
    "ca-cert.pem": "redacted",
    "ca-key.pem": "redacted",
    "server-cert.pem": "redacted",
    "server-key.pem": "redacted"
  },
  "type": "Opaque"
}
Success - secret created.
```

**Comment by mrow4a** [comment]
@liyinan926 what do you think about this change?

File: manifest/spark-operator-with-webhook.yaml [hunk 5]
```diff
@@ -57,6 +57,30 @@ spec:
...
+        app.kubernetes.io/name: sparkoperator
+        app.kubernetes.io/version: v2.3.1-v1alpha1
+    spec:
+      serviceAccountName: sparkoperator
+      restartPolicy: Never
+      containers:
+      - name: main
+        image: gcr.io/spark-operator/spark-operator:v2.3.1-v1alpha1-latest
+        imagePullPolicy: Always
+        command: ["/usr/bin/gencerts.sh", "-n", "sparkoperator", "-p"]
```
Comment (liyinan926) [inline manifest/spark-operator-with-webhook.yaml]: #297 has updated the default namespace used in this repo to `spark-operator` to be consistent with the Helm chart. So the `-n` option can be removed here.

File: manifest/spark-operator-with-webhook.yaml [hunk 6]
```diff
@@ -57,6 +57,30 @@ spec:
...
+        app.kubernetes.io/name: sparkoperator
+        app.kubernetes.io/version: v2.3.1-v1alpha1
+    spec:
+      serviceAccountName: sparkoperator
+      restartPolicy: Never
+      containers:
+      - name: main
+        image: gcr.io/spark-operator/spark-operator:v2.3.1-v1alpha1-latest
+        imagePullPolicy: Always
+        command: ["/usr/bin/gencerts.sh", "-p"]
```
Comment (mrow4a) [inline manifest/spark-operator-with-webhook.yaml]: This requires that the image has newest hack version populated (with new namespace)

**Comment by mrow4a** [comment]
@liyinan926 addressed comments

File: manifest/spark-operator-with-webhook.yaml (hunk 6 above)
Comment (liyinan926) [reply manifest/spark-operator-with-webhook.yaml]: Will do.

**Pull request merged**